src/charm.py

```diff
@@ -401,6 +412,36 @@ def _update_config(self) -> None:
         self._push_config_and_reload(pending)
         self._stored.config_hash = config_hash
 
+    def _get_config(self) -> Union[dict, None]:
+        local_config = self.config["config_file"]
+        remote_config = self.remote_configurer_provider.config()
+        if local_config and remote_config:
+            logger.error("unable to use config from config_file and relation at the same time")
+            raise ConfigUpdateFailure("Multiple configs detected")
+        if local_config:
+            logger.info("using configuration from config_file")
+            return yaml.safe_load(local_config)
+        if remote_config:
+            logger.info("using configuration from relation")
+            return yaml.safe_load(remote_config)  # type: ignore[arg-type]
+        return None
+
+    def _get_templates(self) -> Union[str, None]:
+        local_templates = self.config["templates_file"]
+        remote_templates = self.remote_configurer_provider.templates()
+        if local_templates and remote_templates:
+            logger.error(
+                "unable to use templates from templates_file and relation at the same time"
+            )
+            raise ConfigUpdateFailure("Multiple templates configs detected")
+        if local_templates:
+            logger.info("using templates from templates_file")
+            return local_templates
+        if remote_templates:
+            logger.info("using templates from relation")
```

Comment:
should these logs be debug?

Reply:
Personally, I like when application informs what it does in the logs (on a high level), but I'm open to changing that to debug if that's how you normally handle logging. It's your team's charm after all ;)

Reply:
Doing it once, on `relation_joined`, inside the library is ok. Doing it every time, though... it's a lot of churn.

Reply:
Currently there's no callback implemented for `relation_joined`. 
Do you think it makes sense to implement one just to call `logger` or should I just remove this and assume user knows how does the whole setup work? Or, as suggested by @simskij, change this to `debug` and leave here?

Reply:
Added additional handlers with appropriate loggers.